lib/LANraragi/Controller/Config.pm

```diff
@@ -72,7 +77,12 @@ sub save_config {
         devmode       => ( scalar $self->req->param('devmode')       ? '1' : '0' ),
         enableresize  => ( scalar $self->req->param('enableresize')  ? '1' : '0' ),
         blackliston   => ( scalar $self->req->param('blackliston')   ? '1' : '0' ),
-        nofunmode     => ( scalar $self->req->param('nofunmode')     ? '1' : '0' )
+        tagsruleson   => ( scalar $self->req->param('tagsruleson')   ? '1' : '0' ),
+        nofunmode     => ( scalar $self->req->param('nofunmode')     ? '1' : '0' ),
+
+        # For backend only
+        # Expands rules on save so they can be used directly
+        exptagsrules  => encode_json( [ tags_rules_to_array($self->req->param('tagsrules')) ] )
```

Comment:
Right, I see what you mean now by preferring to store the array in Redis 😁 Storing stuff in ``LRR_CONFIG`` that doesn't directly mirror the configuration entered by the user should be avoided.  

It'd be better to store the computed rule array as a different Redis key directly.  
Redis has rather simple data structures sadly so we can't transpose an array-of-arrays as-is, but I think it's workable with a simple list.  

Here's how this could look like imo(hastily written, probably doesn't run as-is):  

````
# Save
my @array = ("replace_ns","artist","creator","remove","incomplete",null,"strip_ns","test",null);
$redis->del("LRR_TAGRULES"); # Clear previous rules
$redis->lpush("LRR_TAGRULES", reverse @array); # LPUSH always pushes at the head of the list, hence the reverse

# Restore
my @values = $redis->lrange("LRR_TAGRULES", 0,-1);
my @restored = ();

for (my $i=0; $i <= $#values; $i += 3) {
    push @restored, ($values[$i], $values[$i+1], $values[$i+2]);
}
# @restored = ( ("replace_ns","artist","creator"), ("remove","incomplete",null), ("strip_ns","test",null))
````

Reply:
I'll play a bit with this :smile:

Reply:
I added the two subs `save_computed_tagrules` and `get_computed_tagrules` to `Utils/Database`.
You should check them because I think they're not what you expected, but the alternative would have been to make `save_config` and `exec_metadata_plugin` bigger than what they already are...

I also added `flat` from `List::Flatten` without importing the library because it seems that you want to keep the project light.
I had no problem adding test libraries because they can be excluded from the installation, but in this case I prefer to ask before doing that ( you are even using `eval` where I would have used `Try::Tiny` :smile: )